hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/ProvidesHoodieConfig.scala

```diff
@@ -89,6 +89,62 @@ trait ProvidesHoodieConfig extends Logging {
       defaultOpts = defaultOpts, overridingOpts = overridingOpts)
   }
 
+  /**
+   * Get the insert operation.
+   * See if we are able to set bulk insert, else use deduceOperation
+   */
+  private def getOperation(isPartitionedTable: Boolean,
+                           isOverwritePartition: Boolean,
+                           isOverwriteTable: Boolean,
+                           insertModeSet: Boolean,
+                           dropDuplicate: Option[String],
+                           enableBulkInsert: Option[String],
+                           isInsertInto: Boolean,
+                           isNonStrictMode: Boolean,
+                           hasPrecombineColumn: Boolean): String = {
+    val notSetToNonStrict = !insertModeSet || isNonStrictMode
+    //if options are not set, we assume they are configs to do bulk insert
+    (isInsertInto, notSetToNonStrict, enableBulkInsert.getOrElse("true"),
+      dropDuplicate.getOrElse("false"), isOverwritePartition, isPartitionedTable) match {
+      case (true, true, "true", "false", false, _) => BULK_INSERT_OPERATION_OPT_VAL
```

Comment:
for now, can we focus on INSERT_INTO cases w/o bringing in overwrite partition flows. 

we don't need to use the same switch case as well.

we can make this into two sections. 
fetch minimal configs and determine if bulk_insert can be enabled. 
if not, go w/ master code path. 

fetch minimal configs and determine if bulk_insert can be enabled:
```
if InsertInto, If insert mode is not set, bulk insert is not explicitly enabled, dropDuplicate is not set. 
```

In the else section, we can pretty much re-use whats in master. bcoz, for other bulk_insert use-cases there are some guard rails. so, in the above optimization we are adding, lets focus on enabling bulk_insert only when no configs are explicitly set.

Reply:
Consider the case where the user just sets "hoodie.sql.bulk.insert.enable". In your suggestion, we would not end up using bulk insert mode, because the default of "hoodie.sql.insert.mode" is "upsert". Considering that the documentation for the config is "When set to true, the sql insert statement will use bulk insert.", I think that the user is trying to use bulk insert. The way I made it work is that we assume the user wants to use bulk insert until a config is set that is incompatible with bulk insert. In that situation, we then fallback on the original logic